secretHelper.go

```diff
@@ -83,62 +86,93 @@ func (sh *secretHelperImpl) encryptWithKey(unencryptedText, key string, base64en
+func (sh *secretHelperImpl) Decrypt(encryptedTextPlusNonce, pipeline string) (decryptedText, pipelineWhitelist string, err error) {
 
-	return sh.decryptWithKey(encryptedTextPlusNonce, sh.key, sh.base64encodedKey)
+	return sh.decryptWithKey(encryptedTextPlusNonce, pipeline, sh.key, sh.base64encodedKey)
 }
 
-func (sh *secretHelperImpl) decryptWithKey(encryptedTextPlusNonce string, key string, base64encodedKey bool) (decryptedText string, err error) {
+func (sh *secretHelperImpl) decryptWithKey(encryptedTextPlusNonce, pipeline string, key string, base64encodedKey bool) (decryptedText, pipelineWhitelist string, err error) {
+
+	// get decryption key
+	keyBytes, err := sh.getKey(key, base64encodedKey)
+	if err != nil {
+		return
+	}
+	block, err := aes.NewCipher(keyBytes)
+	if err != nil {
+		return
+	}
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return
+	}
 
+	// split string on dots to get nonce, value and pipeline whitelist
 	splittedStrings := strings.Split(encryptedTextPlusNonce, ".")
-	if splittedStrings == nil || len(splittedStrings) != 2 {
+	if splittedStrings == nil || (len(splittedStrings) != 2 && len(splittedStrings) != 3) {
 		err = errors.New("The encrypted text plus nonce doesn't split correctly")
 		return
 	}
 
-	usedNonce := splittedStrings[0]
-	encryptedText := splittedStrings[1]
+	// get nonce
+	nonceBase64 := splittedStrings[0]
+	nonce, _ := base64.URLEncoding.DecodeString(nonceBase64)
 
-	// The key argument should be the AES key, either 16 or 32 bytes to select AES-128 or AES-256.
-	keyBytes, err := sh.getKey(key, base64encodedKey)
+	// get value
+	valueBase64 := splittedStrings[1]
+	valueEncrypted, _ := base64.URLEncoding.DecodeString(valueBase64)
+	valueBytes, err := aesgcm.Open(nil, nonce, valueEncrypted, nil)
 	if err != nil {
 		return
 	}
-	ciphertext, _ := base64.URLEncoding.DecodeString(encryptedText)
+	decryptedText = string(valueBytes)
 
-	nonce, _ := base64.URLEncoding.DecodeString(usedNonce)
+	if len(splittedStrings) == 2 {
+		pipelineWhitelist = DefaultPipelineWhitelist
 
-	block, err := aes.NewCipher(keyBytes)
-	if err != nil {
+		// no need to check pipeline against pipeline whitelist, since the default matches all pipelines
 		return
 	}
 
-	aesgcm, err := cipher.NewGCM(block)
+	// get pipeline whitelist if present
+	pipelineWhitelistBase64 := splittedStrings[2]
+	pipelineWhitelistEncrypted, _ := base64.URLEncoding.DecodeString(pipelineWhitelistBase64)
+	pipelineWhitelistBytes, err := aesgcm.Open(nil, nonce, pipelineWhitelistEncrypted, nil)
 	if err != nil {
 		return
 	}
+	pipelineWhitelist = string(pipelineWhitelistBytes)
 
-	plaintext, err := aesgcm.Open(nil, nonce, ciphertext, nil)
+	// check if pipeline is matched by pipeline whitelist regular expression
+	pattern := fmt.Sprintf("^%v$", pipelineWhitelist)
+	validForPipeline, err := regexp.MatchString(pattern, pipeline)
 	if err != nil {
 		return
 	}
-
-	decryptedText = string(plaintext)
+	if !validForPipeline {
```

Comment:
Would be nice if this check was made before calling the `Open` method on the actual secret. But not a big deal.

Reply:
It returns an empty value it it fails, but i can swap them. It makes harder to return early though.

Reply:
Yeah, I guess if someone manages to hack this by somehow getting the memory value before this fails, we would be already doomed in so many other ways :joy: 
So I think it's good as it is.